daemon/src/model/cfd.rs

```diff
@@ -363,6 +365,221 @@ impl Cfd {
+
+                    MustRefund {
+                        common: CfdStateCommon {
+                            transition_timestamp: SystemTime::now(),
+                        },
+                        dlc,
+                    }
+                }
+                monitor::Event::RefundFinality(_) => {
+                    match self.state {
+                        MustRefund { .. } => (),
+                        OutgoingOrderRequest { .. } => unreachable!("taker-only state"),
+                        IncomingOrderRequest { .. }
+                        | Accepted { .. }
+                        | Rejected { .. }
+                        | ContractSetup { .. } => {
+                            bail!("Did not expect refund finality yet: ignoring")
+                        }
+                        PendingOpen { .. } => {
+                            tracing::debug!(%order_id, "Was waiting on lock finality, jumping ahead");
+                        }
+                        Open { .. } => {
+                            tracing::debug!(%order_id, "Was waiting on CET timelock expiry, jumping ahead");
+                        }
+                        OpenCommitted { .. } => {
+                            tracing::debug!(%order_id, "Was waiting on refund timelock expiry, jumping ahead");
+                        }
+                        Refunded { .. } => bail!("State already assumes refund finality: ignoring"),
+                    }
+
+                    Refunded {
+                        common: CfdStateCommon {
+                            transition_timestamp: SystemTime::now(),
+                        },
+                    }
+                }
+            },
+        };
+
+        Ok(new_state)
+    }
+
+    pub fn refund_tx(&self) -> Result<Transaction> {
+        let dlc = if let CfdState::MustRefund { dlc, .. } = self.state.clone() {
+            dlc
+        } else {
+            bail!("Refund transaction can only be constructed when in state MustRefund, but we are currently in {}", self.state.clone())
+        };
+
+        let sig_hash = spending_tx_sighash(
+            &dlc.refund.0,
+            &dlc.commit.2,
+            Amount::from_sat(dlc.commit.0.output[0].value),
+        );
+        let our_sig = SECP256K1.sign(&sig_hash, &dlc.identity);
+        let our_pubkey = PublicKey::new(bdk::bitcoin::secp256k1::PublicKey::from_secret_key(
+            SECP256K1,
+            &dlc.identity,
+        ));
+        let counterparty_sig = dlc.refund.1;
+        let counterparty_pubkey = dlc.identity_counterparty;
+        let signed_refund_tx = finalize_spend_transaction(
+            dlc.refund.0,
+            &dlc.commit.2,
+            (our_pubkey, our_sig),
+            (counterparty_pubkey, counterparty_sig),
+        )?;
+
+        Ok(signed_refund_tx)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum CfdStateChangeEvent {
+    // TODO: groupd other events by actors into enums and add them here so we can bundle all
```

Comment:
:question: Is this TODO up-to-date?

Reply:
It is meant for all other state transitions (outside of monitoring). the idea is to bundle *all* state transitions of a Cfd in this function. 
That said, it is still valid (but has a typo...)